### ISSUE TITLE
Request: In FtpClient, make UploadFileInternal and DownloadFileInternal virtual.

### ISSUE THREAD

**Issue by russjudge** [robinrodricks/FluentFTP#1684]
By making UploadFileInternal and DownloadFileInternal virtual, a class that inherits FtpClient can override these methods.  The reason I want this is to add code that automatically adjusts the UploadRateLimit and DownloadRateLimit in the Config object based on detected transfer speeds.  My reason for building this code is to solve a problem whereby an upload always failed until through trial-and-error I found a rate limit that worked.  When I uploaded the same file through a standard client (FileZilla), the upload succeeded 100% of the time.

Since UploadFileInternal and DownloadFileInternal are not virtual, my class that inherits FtpClient has to hide the public methods that call these methods to use my alternative, which wraps UploadFileInternal and DownloadFileInternal to add transfer rate controlling.  Making these methods virtual will simplify the code for my class that inherits FtpClient.

**Comment by FanDjango** [comment]
> solve a problem whereby an upload always failed until...

Just curious: Failed how?

Instead of rate limiting, as compared to standard (FileZilla? is it such, really?), I would be WireSharking the heck out of this first.

**Comment by robinrodricks** [comment]
This has been fixed in the latest release: https://www.nuget.org/packages/FluentFTP/52.0.0

**Comment by FanDjango** [comment]
Thanks for this additional information.

> When I tried FileZilla to reproduce the error outside of FluentFTP, FileZilla had no issue.

This always hurts a little and it would be nice to find the reason. That was my motivation, is all...

Was there something subtly pointing you to, you know, think tinkering with the rate (of all things) would seem a good idea? 

Anyway, if rate limiting solves your problem in some mysterious way, that's fine.

Hoping that the problem does not reappear at the most inopportune time.

**Comment by russjudge** [comment]
> Thanks for this additional information.
> 
> > When I tried FileZilla to reproduce the error outside of FluentFTP, FileZilla had no issue.
> 
> This always hurts a little and it would be nice to find the reason. That was my motivation, is all...
> 
> Was there something subtly pointing you to, you know, think tinkering with the rate (of all things) would seem a good idea?
> 
> Anyway, if rate limiting solves your problem in some mysterious way, that's fine.
> 
> Hoping that the problem does not reappear at the most inopportune time.

Yeah, at the point of tinkering with rate limiting I was desparate to simply get my process to work, and I was trying anything.  The difference between FileZilla and FluentFTP seems to make no sense.  I can tell you a few other things that might help to make a little sense--the communication is entirely internal to the company I work for, and the file it tripped on was a .dll.  Today I corrupted that dll by simply putting some random data into the file, and it no longer tripped on it.  I am starting to suspect that the issue may be linked to some kind of anti-virus software running on the server, though why FileZilla has no trouble makes no sense to me.  Speculating, it seems that by slowing the rate of transfer, the anti-virus had more time to analyze the data.  One symptom I came across is that with rate limiting, if I did not limit sufficiently, multiple quick uploads of the same file would eventually cause the error, and then for some time the error would occur more quickly, unless I restarted the FTP server.

**Comment by FanDjango** [comment]
VERY interesting, thanks for this. I actually beginning to enjoy myself here and hope that you don't mind.

> Today I corrupted that dll by simply putting some random data into the file, and it no longer tripped on it

Very astute of you to try that. Obviously, your are most familiar with your own environment, so you are the best one to seek an explanation.

> The difference between FileZilla and FluentFTP seems to make no sense

If you assume that *the phyiscal contents, the data itself* that is being transmitted might be triggering a content aware firewall or router, or virus scanner on the server, note that FileZilla encrypts with GnuTLs, whichever TLS version is negotiated, whichever cypher suite is chosen, and I assume your FluentFTP is currently using .NET SslStream. So in the end the data on the line differs. At the receving end, of course, the data should again be equivalent. So it looks like the transfer process itself is ok.

If the server shows no error and the transmission is simply terminated "magically", wireshark would show an abrupt RST from the server side, but it would not "prove" much beyond your theory, it would just corroborate it.

> some kind of anti-virus software running on the server, though why FileZilla has no trouble makes no sense to me

I agree. But someone might have configured an exemption for the FileZilla executable. If it is a virus protection scheme, a professional one, then perhaps you will have been triggering security breaches that are logged somewhere in the endpoint management solution.

So, you have convinced me about what you are up against. Your risk remains: Some other transfer might exhibit similar problems and might not be solvable with a rate reduction.

Don't hesitate to re-open, or reach out when you need more.

**Comment by russjudge** [comment]
> VERY interesting, thanks for this. I am actually beginning to enjoy myself here and hope that you don't mind.
> 
> > Today I corrupted that dll by simply putting some random data into the file, and it no longer tripped on it
> 
> Very astute of you to try that. Obviously, your are most familiar with your own environment, so you are the best one to seek an explanation.
> 
> > The difference between FileZilla and FluentFTP seems to make no sense
> 
> If you assume that _the phyiscal contents, the data itself_ that is being transmitted might be triggering a content aware firewall or router, or virus scanner on the server, note that FileZilla encrypts with GnuTLs, whichever TLS version is negotiated, whichever cypher suite is chosen, and I assume your FluentFTP is currently using .NET SslStream. So in the end the data on the line differs. At the receving end, of course, the data should again be equivalent. So it looks like the transfer process itself is ok.
> 
> If the server shows no error and the transmission is simply terminated "magically", wireshark would show an abrupt RST from the server side, but it would not "prove" much beyond your theory, it would just corroborate it. But the RST scenario mostly points to a CA-firewall or a CA-Router. Virus intervention looks different - file transer successful but file disappears, can't be found.
> 
> > some kind of anti-virus software running on the server, though why FileZilla has no trouble makes no sense to me
> 
> I agree. But someone might have configured an exemption for the FileZilla executable. If it is a virus protection scheme, a professional one, then perhaps you will have been triggering security breaches that are logged somewhere in the endpoint management solution.
> 
> So, you have convinced me about what you are up against. Your risk remains: Some other transfer might exhibit similar problems and might not be solvable with a rate reduction.
> 
> Don't hesitate to re-open, or reach out when you need more.

As an FYI--this is over straight FTP--not using any kind of encryption that would need TLS or anything like that.  Our virus protection is managed by a third-party, and I don't know the details (beyond the scope of my concern).  The server is a virtual server that I configured myself after it was initially built by the third-party.

I'm happy to provide this information--glad you enjoyed it.

**Comment by FanDjango** [comment]
> this is over straight FTP

This is really strange, then.
